tests/flytekit/unit/core/test_unions.py

```diff
@@ -0,0 +1,93 @@
+import typing
+from dataclasses import dataclass
+from enum import Enum
+import sys
+import pytest
+
+from flytekit.core.context_manager import FlyteContextManager
+from flytekit.core.type_engine import TypeEngine, TypeTransformerFailedError
+
+
+def test_asserting():
+    @dataclass
+    class A:
+        a: str = None
+
+    @dataclass
+    class B:
+        b: str = None
+
+    @dataclass
+    class C:
+        c: str = None
+
+    ctx = FlyteContextManager.current_context()
+
+    pt = typing.Union[A, B, str]
+    lt = TypeEngine.to_literal_type(pt)
+    # mimic a register/remote fetch
+    guessed = TypeEngine.guess_python_type(lt)
+
+    TypeEngine.to_literal(ctx, A("a"), guessed, lt)
+    TypeEngine.to_literal(ctx, B(b="bb"), guessed, lt)
+    TypeEngine.to_literal(ctx, "hello", guessed, lt)
+
+    with pytest.raises(TypeTransformerFailedError):
+        TypeEngine.to_literal(ctx, C("cc"), guessed, lt)
+
+    with pytest.raises(TypeTransformerFailedError):
+        TypeEngine.to_literal(ctx, 3, guessed, lt)
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 10), reason="enum checking only works in 3.10+"
+)
+def test_asserting_enum():
+    class Color(Enum):
+        RED = "one"
+        GREEN = "two"
+        BLUE = "blue"
+
+    lt = TypeEngine.to_literal_type(Color)
+    guessed = TypeEngine.guess_python_type(lt)
+    tf = TypeEngine.get_transformer(guessed)
+    tf.assert_type(guessed, "one")
+    tf.assert_type(guessed, guessed("two"))
+    tf.assert_type(Color, "one")
+
+    guessed2 = TypeEngine.guess_python_type(lt)
+    tf.assert_type(guessed, guessed2("two"))
+
+
+@pytest.mark.skipif(
+    sys.version_info >= (3, 10), reason="3.9 enum testing"
+)
+def test_asserting_enum_39():
+    class Color(Enum):
+        RED = "one"
+        GREEN = "two"
+        BLUE = "blue"
+
+    lt = TypeEngine.to_literal_type(Color)
+    guessed = TypeEngine.guess_python_type(lt)
+    tf = TypeEngine.get_transformer(guessed)
+    tf.assert_type(guessed, guessed("two"))
+    tf.assert_type(Color, Color.GREEN)
+
+
+@pytest.mark.sandbox_test
+def test_with_remote():
```

Comment:
Do we have to make this test run in CI? I don't think we're running any of the tests marked as `sandbox_test` currently.

Reply:
no this is just to have for local testing